pkg/services/ngalert/models/admin_configuration.go

```diff
@@ -34,3 +45,7 @@ func (ac *AdminConfiguration) Validate() error {
 
 	return nil
 }
+
+func (amc AlertmanagersChoice) IsValid() bool {
+	return amc == AllAlertmanagers || amc == InternalAlertmanager || amc == ExternalAlertmanagers
```

Comment:
(this is perhaps more of a question than a comment)

We're mapping this Swagger enum to a plain string in Go. I didn't notice anything that explicitly sets `all` to be the default value of `AlertmanagersChoice` if this key isn't present in the request. Does the automatic Go <-> Swagger binding do this for us since `all` is the zeroth listed value in the Swagger doc? What happens if a request without this key is sent?

I ask because Swagger generators/bindings for other langs that I've used in the past haven't typically done this automatically, and will usually either map missing values to the default of the underlying type, or reject the request entirely. If the Go integration works this way, does this method reject requests without the key, causing a backward compat break?

Reply:
There's no conversion in place for missing values. If an empty value is sent, a `400` error is returned. The `alertmanagersChoice` parameter is required.

Converting missing values into one of the three options would be a breaking change since we're assuming what the user meant to send, and potentially changing the behavior of the alerting system.

For example, let's say a user has their system set up so that all alerts are sent to the external Alertmanagers. If the user sends a request without a value for `alertmanagersChoice`, that value would default to `AllAlertmanagers`, changing the current behavior of the system and thus resulting in all alerts being sent to both the internal and external Alertmanagers.

Reply:
This way is breaking too, though. In the example above, the user's system would simply stop working altogether after upgrade, because it no longer knows how to send valid requests.

At what granularity is it acceptable to submit an API-breaking change? If we follow semantic versioning strictly, we can't break compat until the next major version release. If we're moving toward a multi-tenant and fully stateless Grafana, we'll eventually need to get comfortable with *never* breaking APIs and using API versioning instead.

If we're okay with breaks on a minor release for now I'm good with it, just wanting to be consistent as to what our strategy is for this.

Reply:
FWIW I think this is a great change overall. :+1: Just concerned about the edge case where users are upgrading. If `AllAlertmanagers` isn't how the system works today, do we think it'd be reasonable to add a fourth enum option that represents exactly the behavior that clients see today? If someone wants the current behavior and it's unrepresentable in the new API, I could see them being frustrated.

Reply:
@alexweav you have a great point! I don't think adding a fourth enum option would be useful, since they would still have to provide some value for the `alertmanagersChoice` field.

I can try to make so that if there's no value for `alertmanagersChoice`, then no change is applied to that part of the admin configuration. This would be backwards compatible. Also, I don't think there's a current behavior that is unrepresentable by the new API, so I think I'll go this way.

Reply:
After some thought I decided that defaulting to `AllAlertmanagers` is the right choice since it's how the system is working currently.